### PR TITLE
8342826: Improve performance of oopDesc::klass() after JDK-8305895

### PR DESCRIPTION
**Pull request by rkennke**
Before intergration of JEP 450, a number of minor regressions have been identified. The root-cause of those is the additional flag-check for UseCompactObjectHeaders in a couple of hot code-paths. This change addresses those cases by introducing a new helper class ObjLayout, which initializes some global state depending on the flags, and use that state later, instead of loading and checking multiple flags in hot paths.

This solution is not great. The real fix will eventually be to get rid of UseCompressedClassPointers in a first step, and later also get rid of UseCompactObjectHeaders, and settle on a single object layout. But we are not there, yet, and it will take several (or many) releases to get there. In the meantime, the proposed change eliminates the remaining known regressions.

Relevant benchmarks:

DaCapo:pmd (less is better)
pre-jep450:  703.67ms
mainline:  729.38ms
alwaysinline: 730.45ms
jdk8342826:  704.25ms

CryptoRsa (more is better)
pre-jep450: 9315.719 ops/min
mainline: 10109.509 ops/min
alwaysinline: 10232.120
jdk8342826: 10272.161 ops/min
jdk8342826+alwaysinline: 10072.193 ops/min

Throw.throwWith64Frames microbenchmark
Before JEP 450 (605396280d5ea225828da4ed688068334a15e122)
Throw.throwWith64Frames  avgt   40  3943.690 ± 15.456  ns/op
Mainline
Throw.throwWith64Frames  avgt   40  4083.029 ± 12.044  ns/op
JDK-8342826
Throw.throwWith64Frames  avgt   40  3973.082 ± 12.956  ns/op

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342826](https://bugs.openjdk.org/browse/JDK-8342826): Improve performance of oopDesc::klass() after JDK-8305895 (**Bug** - P3)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**) Review applies to [ee7be903](https://git.openjdk.org/jdk/pull/22020/files/ee7be903595fd5204ceed90075e22d343059afbb)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22020/head:pull/22020` \
`$ git checkout pull/22020`

Update a local copy of the PR: \
`$ git checkout pull/22020` \
`$ git pull https://git.openjdk.org/jdk.git pull/22020/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22020`

View PR using the GUI difftool: \
`$ git pr show -t 22020`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22020.diff">https://git.openjdk.org/jdk/pull/22020.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22020#issuecomment-2474781514)
</details>

### CODE REVIEW

**Comment by stefank** [comment]
Could you make an experiment where you just force-inline `oopDesc::klass()` with `ALWAYSINLINE` and compare that to the proposed patch? When I investigated one regression a non-inlined `klass()` was prominent in the JEP-450 code but not in the baseline.

**Comment by rkennke** [comment]
> Could you make an experiment where you just force-inline `oopDesc::klass()` with `ALWAYSINLINE` and compare that to the proposed patch? When I investigated one regression a non-inlined `klass()` was prominent in the JEP-450 code but not in the baseline.

Force-inlining doesn't help at all. dacapo:pmd scores at 737.51ms, which is about what I currently see in mainline.

**Review by coleenp — COMMENTED** [review]
I have sort of a different idea.  Instead of these switch statements, can you templatize these functions based on ObjLayoutMode?

**Comment by rkennke** [comment]
> I have sort of a different idea. Instead of these switch statements, can you templatize these functions based on ObjLayoutMode?

There are a few places where these methods (or callers of these methods, like oopDesc::size()) are used in hot loops. In those cases I guess we could do that. But templatizing the methods always needs 'driver methods', e.g. some place that decides, based on the runtime value of the flags, which temlatized loop to call. It's ugly.

AFAIK, @tstuefe has done something similar and slightly better than that in his KLUTE prototype, but it's also more complex and nothing that helps us right now with the problems that we have (e.g. in stack-trace building, which is what affected us in dacapo:pmd).

**Review by coleenp — COMMENTED** [review]
I guess I retract that suggestion as it probably wouldn't help.  And now I see your reply.

**Review by coleenp — APPROVED** [review]
Okay, this looks good.  Testing one global flag rather than two seems an improvement.

**Review by Hamlin-Li — APPROVED** [review]
Looks good!

File: src/hotspot/share/oops/objLayout.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *   ..
+ * }
+ */
+class ObjLayout {
+public:
+  enum Mode {
+    // +UseCompactObjectHeaders (implies +UseCompressedClassPointers)
+    Compact,
+    // +UseCompressedClassPointers (-UseCompactObjectHeaders)
+    Compressed,
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/oops/objLayout.hpp:57]: Hilariously, it looks counter-productive, at least for my GCC builds. `Compressed` now does two jumps in trivial method that just calls to `oopDesc::klass()`.

`Compact = 0`, `Compressed = 1`:

```
Klass* oopDesc::klass_non_inlined() {
 ab7:   55                      push   %rbp
 ab8:   48 89 e5                mov    %rsp,%rbp
  switch (ObjLayout::klass_mode()) {
 abb:   8b 00                   mov    (%rax),%eax
 abd:   85 c0                   test   %eax,%eax
 abf:   74 26                   je     ae7
 ac1:   83 f8 01                cmp    $0x1,%eax
 ac4:   75 1b                   jne    ae1
 ac6:   8b 47 08                mov    0x8(%rdi),%eax  ; <--- Compressed
 ac9:   48 8b 15 00 00 00 00    mov    0x0(%rip),%rdx     
   return klass();
}
 ad0:   5d                      pop    %rbp
 ad1:   8b 0a                   mov    (%rdx),%ecx
 ad3:   48 8b 15 00 00 00 00    mov    0x0(%rip),%rdx 
 ada:   48 d3 e0                shl    %cl,%rax
 add:   48 03 02                add    (%rdx),%rax
 ae0:   c3                      retq
      return _metadata._klass;
 ae1:   48 8b 47 08             mov    0x8(%rdi),%rax ; <--- Uncompressed
 ae5:   5d                      pop    %rbp
 ae6:   c3                      retq
    return *dest;
 ae7:   48 8b 07                mov    (%rdi),%rax
 aea:   48 c1 e8 2a             shr    $0x2a,%rax  ; <--- Compact
 aee:   eb d9                   jmp    ac9
``` 

`Compressed = 0`, `Compact= 1`:

```
Klass* oopDesc::klass_non_inlined() {  
 ab7:   55                      push   %rbp
 ab8:   48 89 e5                mov    %rsp,%rbp
  switch (ObjLayout::klass_mode()) {
 abb:   8b 00                   mov    (%rax),%eax
 abd:   85 c0                   test   %eax,%eax
 abf:   74 2a                   je     aeb
 ac1:   83 f8 01                cmp    $0x1,%eax
 ac4:   75 1f                   jne    ae5
    return *dest;
 ac6:   48 8b 07                mov    (%rdi),%rax
 ac9:   48 c1 e8 2a             shr    $0x2a,%rax      ; <--- Compact
 acd:   48 8b 15 00 00 00 00    mov    0x0(%rip),%rdx   
   return klass();
}
 ad4:   5d                      pop    %rbp
 ad5:   8b 0a                   mov    (%rdx),%ecx
 ad7:   48 8b 15 00 00 00 00    mov    0x0(%rip),%rdx
 ade:   48 d3 e0                shl    %cl,%rax
 ae1:   48 03 02                add    (%rdx),%rax
 ae4:   c3                      retq
      return _metadata._klass;  
 ae5:   48 8b 47 08             mov    0x8(%rdi),%rax  ; <--- Uncompressed
 ae9:   5d                      pop    %rbp
 aea:   c3                      retq   
 aeb:   8b 47 08                mov    0x8(%rdi),%eax  ; <--- Compressed
 aee:   eb dd                   jmp    acd

```

**Comment by shipilev** [comment]
Pull from fresh master to get macos-aarch64 builds fixed.

File: src/hotspot/share/oops/objLayout.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/oops/objLayout.hpp:57]: Ok, that's surprising to me.  You can change it back.

**Comment by stefank** [comment]
> > Could you make an experiment where you just force-inline `oopDesc::klass()` with `ALWAYSINLINE` and compare that to the proposed patch? When I investigated one regression a non-inlined `klass()` was prominent in the JEP-450 code but not in the baseline.
> 
> Force-inlining doesn't help at all. dacapo:pmd scores at 737.51ms, which is about what I currently see in mainline.

OK. Funny enough. When I force inlined klass() and the backtrace push function I got back the pmd scores. Still, for future reference can you add the benchmark scores for all the listed benchmarks for 'mainline with ALWAYSINLINE klass()'?

**Comment by rkennke** [comment]
> > > Could you make an experiment where you just force-inline `oopDesc::klass()` with `ALWAYSINLINE` and compare that to the proposed patch? When I investigated one regression a non-inlined `klass()` was prominent in the JEP-450 code but not in the baseline.
> > 
> > 
> > Force-inlining doesn't help at all. dacapo:pmd scores at 737.51ms, which is about what I currently see in mainline.
> 
> OK. Funny enough. When I force inlined klass() and the backtrace push function I got back the pmd scores. Still, for future reference can you add the benchmark scores for all the listed benchmarks for 'mainline with ALWAYSINLINE klass()'?

It *does* seem to help with the Crypto benchmark, though. Also, I am not sure why, but when I tested JEP450, I could see a very reproducible regression with JEP450, but current benchmark shows quite an improvement, instead. Something else must have changed since then.

**Comment by stefank** [comment]
FWIW, I ran some experiments with CryptoRSA and:
```
java  -server -Xmx4g -Xms4g -XX:+AlwaysPreTouch -Xlog:gc -jar ./SPECjvm2008.jar --showversion crypto.rsa -it 60 -wt 30 -ikv
```

mainline:
8820.19 ops/m

PR (94275040e36 Reorder 'Compact' vs 'Compressed'):
9679.77 ops/m

PR but skip 94275040e36:
9921.99 ops/m

mainline inline klass():
9594.34 ops/m

mainline inline klass() + restructure klass()
10032.58 ops/m

The last experiment was with the odd-looking
```
@@ -95,12 +95,14 @@ void oopDesc::init_mark() {
 }

 Klass* oopDesc::klass() const {
-  if (UseCompactObjectHeaders) {
-    return mark().klass();
-  } else if (UseCompressedClassPointers) {
-     return CompressedKlassPointers::decode_not_null(_metadata._compressed_klass);
+  if (!UseCompactObjectHeaders) {
+    if (!UseCompressedClassPointers) {
+      return _metadata._klass;
+    } else {
+      return CompressedKlassPointers::decode_not_null(_metadata._compressed_klass);
+    }
   } else {
-    return _metadata._klass;
+    return mark().klass();
   }
 }
```
I wanted to test if messing around with the order here would give me similar performance differences that you were seeing when you changed the order in the PR.

**Comment by rkennke** [comment]
I confirm that using the odd reorder does indeed lead to equally good performance in the cryptorsa benchmark. However, I also tried the same reordering in base_offset_in_bytes() and it has adverse effects there. Overall, this approach doesn't seem very sustainable: we don't know how other compilers or even just other versions of the same compiler behave similar, and even then we don't know how that works out on other CPUs (or other versions of the same CPU category). I also played with __builtin_expect, and while it was useful, it did not lead to better performance than this PR.

Overall, I think switching over a single global variable seems the sanest approach and most likely to work well across compilers and CPUs. In the long run, we should prioritize to get rid of the flags ASAP (but it will surely take a couple of release cycles to get there).

Are we good to go with the current version of this PR?

**Review by shipilev — APPROVED** [review]
Yeah, I am good.

**Review by coleenp — APPROVED** [review]
Looks good to me.  Thanks for the extra performance testing.

**Comment by rkennke** [comment]
Thanks!
/integrate

**Pull request closed without merging**